### ISSUE TITLE
Accessibility issues

### ISSUE THREAD

**Issue by peterfabian** [woocommerce/woocommerce#24185]
ref: https://woocommerce.com/posts/building-for-web-accessibility/#comment-9029518

---

:bulb: If you have identified an additional accessibility issue, please log it in a [new issue](https://github.com/woocommerce/woocommerce/issues/new/choose) rather than appending it  here. This makes it easier for us to assess and for us to coordinate efforts. Thank you!

**Comment by vadimhp** [comment]
Agreed.

Also on the cart page: the "Update" button doesn't need to be in the row of the products table. That makes the table used for layout which it shouldn't be. The button can just be in its own `div` following the table.

Additionally, updating cart items quantities doesn't inform screen readers of any activity/results triggered by the "Update" button.

Similarly, the ajax-add-to-cart buttons on other pages do not inform screen readers of any activity/updates about item additions to cart.

To address the latter two issues, I suggest creating a `<div role=status class=screen-reader-text>` after each one of these buttons (the add-to-cart and the cart-update buttons) and use events such as `adding_to_cart`, `added_to_cart`, and `cart_page_refreshed` to populate that div dynamically.

**Comment by vanessasickles** [comment]
Some additional issues that I've found frequently on our installs using the [WAVE Evaluation Tool](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh?hl=en-US):
- The coupon code input doesn't have a label
- The quantity input's "arialabelledby" doesn't have content, throwing an error
- The sort/filter/orderby on product rolls doesn't have a label

**Comment by reidelliott** [comment]
The `$subtext` SCSS variable, which is used on `.breadcrumb` is `#777`, which, if on white renders color contrast errors. Darkening once to `#767676` would solve this.

**Comment by coreybullman** [comment]
We recently had a 3rd-party company perform an accessibility audit on one of our WooCommerce shops. The following issues came out of the audit.

### Cart page product table
The "X" remove links do not have meaningful alternative text. The aria-label “Remove this item” isn’t meaningful for customers using a screenreader. Adding something like `aria-describedby=""` that matched a corresponding ID for the product link in the cart table would provide better context.

### Single product pages
The headings for the products in the "Related Products" section use an H2 tag but don’t really designate a new section so the logical reading order for screenreaders is thrown off. It was suggested to update the title elements to use a different element that doesn’t designate a new section (e.g. `<div>`).

### Checkout page
**[Issue]** The "State" `<select>` field has an aria-hidden true attribute applied which removes it from the screen reader reading order. Instead, there is a `<span>` with `role="combobox"` that the user will interact with. This control does not have an accessible name as the `aria-labelledby` is referencing the ID of a nested <span> which has a `role="textbox"`.

**[User Impact]** When controls do not provide name, role and/or state, screen reader users will not know their purpose and current state (if applicable).

### Cart page
**[Issue]** The "Update Cart" button does not need an aria-disabled attribute, as it has a disabled attribute.

**[User Impact]** When ARIA attributes (state, roles, and properties) are not valid roles according to the specification or are not used correctly, assistive technology may not function as expected.

### Product list views
**[Issue]** The following link elements do not have meaningful text:

- Select Options
- Add to Cart

**[User Impact]** Without meaningful text, screen reader users who navigate the page or navigate using a list of links (provided by their Assistive Technology, e.g., JAWS, VoiceOver, etc...) will not be able to discern the purpose of these links.

**[Potential fix]** Add an `aria-describedby=""` attribute to the buttons that references of the corresponding product.

### Cart and Checkout pages
**[Issue]** Errors are not programmatically associated or within proximity of their fields.

**[User Impact]** Users of screen readers will have a difficult time understanding why a form is being rejected, which fields are in error, and how to remedy them.

**[Recommendation]** Ensure `aria-invalid="true"` is on invalid form fields and updated to `aria-invalid="false"` or removed for valid fields.

- Ensure error messages are nested inside of the <label> element.
- Ensure errors that can't be nested with the <label> element are programmatically associated with the input. Add aria-describedby to the input element and point to the
- ID of the error message.
- Ensure focus moves to the first field that has an error.

**Comment by helgatheviking** [comment]
I also have been looking at accessibility for my own plugins and in the audit they mentioned several things about my content that are generated by WC. While I am trying to get better at it, this isn't my specialty so relaying what was reported to me: 

# Single Product Page

## figure-element
The `<figure>` around the link is an unwanted element. WCAG doesn’t mention this as a violation and according to the HTML specs, this shouldn’t be an issue. However, a screen reader user will run into problems. 

a screen reader will read:
“Clickable figure clickable link image img-ratione.png ..."
“Clickable” is read twice for one link. 
There has been a consultation with other A11y experts on this topic to learn whether this is a violation, and it’s not. But it’s not a very elegant solution.

Suggestion: if possible, remove `<figure>` and replace with a meaningless HTML-element like `<div>`

## Title attribute
Best practice: On the quantity input field, the title attribute is filled out. I advice you not to use the title attribute. This sometimes gets in the way on hover. The title attribute sometimes is and sometimes isn’t read by screen readers. You can only use this attribute on hover, so it requires a mouse. People who do not use a mouse (eg. on tablet or mobile), cannot use this attribute. On zoom, all text becomes really large, except for the tooltip on the title attribute. See the W3C HTML 5 recommendation: https://html.spec.whatwg.org/multipage/dom.html#the-title-attribute

## Label next to Add to cart button
Next to the add to cart button there’s an input field. The visual label is missing and therefore it’s not clear what the number does/means. Show a label (WCAG).

(from me: I don't think I've ever seen a quantity label string on any ecommerce shop! but relaying the message)

# Shop Page

## Duplicate links
People who can see, but who depend on keyboard navigation, need to use the tab key for every product. It’s a best practice to have only one link for the product container. (the product image and the product title both have links)

**Comment by aurorahawaii** [comment]
We've tested Woocommerce with NVDA screen reader and Wave accessibility tool, and noticed some accessibility errors. Here is a brief summary:

1. **Checkout Page** – Coupon Code field uses placeholder text for a label. Provide visible and persistent labels for form controls (i.e. using the label element).
2. **Checkout Page** – Gateway icon group for PayPal checkout is missing alternate text (i.e. PayPal Credit, PayPal, Mastercard, etc).
3. **Shop Page** – Product sort option triggers a change of context without warning users. Add a button to allow users to trigger changes in context.

![Woocommerce-Coupon-Code-Label](https://user-images.githubusercontent.com/87996570/127032141-9dc30691-406f-40d6-a65a-66d6a6eecddb.jpg)
![Woocommerce-Payment-Gateways](https://user-images.githubusercontent.com/87996570/127032156-5d4cc3d3-37ef-4a5e-99d5-64befd1252c3.jpg)
![Woocommerce-Product-Sort](https://user-images.githubusercontent.com/87996570/127032358-6700ad2d-8ce8-4e04-918c-7c1220a9aa30.jpg)

Thanks!

**Comment by thuautp** [comment]
From 4354659-zen, related to accessibility:

> Our website is scanned on accessbility WCAG 2.1 norms by an expert. Some of the issues are related to our WooCommerce setup: 1) The messages lack a clear instruction: e.g. Billing Street address is a required field -> better, The field Billing Street address is not filled, please fill in your Billing Street address. 2) The location of the error message is not helpful: instead of putting it on top, it's better to position the message next to the required field.

**Comment by peterfabian** [comment]
Thank you all for your valuable input here. I'm increasing the priority of this issue so that we start prioritizing and fixing these issues soon.

**Comment by betayazilim** [comment]
Hi,
there is "ARIA input fields do not have accessible names" error in product filters.

Woocommerce widget as “Filter products by attributes” and select display type as “Dropdown” this error happens.
Any advice to fix it?

Error screenshot : https://prnt.sc/26ghisp
Details about aria name : https://web.dev/aria-name/
Woocommerce : 6.1.0
WordPress : 5.8.3

**Comment by betayazilim** [comment]
Another one.
Woocommerce additional product images don't have alt tag in slider.

Image elements do not have [alt] attributes

Error : https://prnt.sc/26h04dr , https://prnt.sc/26h054q
Demo : https://themes.woocommerce.com/storefront/product/build-your-dslr/

**Comment by gslwebsolutions** [comment]
Shipping radio options lose focus onchange. 
When I tab to the shipping options, I am able to switch choices using the up and down arrow keys. However, after switching to a different option and Ajax loads, I seem to lose keyboard focus and I am now unable to use the up and down arrow keys. I have to press Tab again to be able to use the arrow keys again.

**Comment by stronenv** [comment]
I don't know if this issue is meant to be a running tally of a11y issues, or if you would like me to open a new issue?

But we're seeing an issue with the "Filter products by attribute" on the Product List Views as [mentioned above](https://github.com/woocommerce/woocommerce/issues/24185#issuecomment-1018669224).

The select element does not have an associated label.

**Comment by barryhughes** [comment]
With [Contributor Day](https://developer.woocommerce.com/2023/03/31/woocommerce-contributor-day/) happening soon, we've taken a moment to break most of the noted problems into atomic issues of their own—this will make coordination a lot easier, and should help prevent duplicated efforts. If your issue has not made it into a new issue, it could be because:

- The problem has already been resolved.
- It belongs in a sister project, such as [WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks).
- We could not identify the problem based on the information provided.

Or, we may have made a mistake. If you feel there is a still a problem to be addressed, though, simply create a **[new issue](https://github.com/woocommerce/woocommerce/issues/new/choose)** and we'll be happy to take a look. Thanks for all the ideas!
